gap/SnAnUnknownDegree.gi

```diff
@@ -0,0 +1,713 @@
+        local
+            # integer, loop variable
+            a,
+            # elements, in G
+            r, tPower, tPowerOld, c;
+        # Steps 2 & 3: New involution
+        # Check if we either tried enough conjugates or constructed enough
+        # three cycle candidates for the current involution t.
+        # If this is the case, we need to construct the next involution
+        if nrTriedConjugates >= C or nrThreeCycleCandidates >= T then
+            r := RandomElm(ri, "simplesocle", true)!.el;
+            a := 0;
+            tPower := r ^ M;
+            # Invariant: tPower = (r ^ M) ^ (2 ^ a)
+            repeat
+                a := a + 1;
+                tPowerOld := tPower;
+                tPower := tPower ^ 2;
+            until a = logInt2N or isone(ri)(tPower);
+            if a = logInt2N then
+                return NeverApplicable;
+            fi;
+            t := tPowerOld;
+            nrInvolutions := nrInvolutions + 1;
+            nrTriedConjugates := 0;
+            nrThreeCycleCandidates := 0;
+        fi;
+        # Steps 4 & 5: new three cycle candidate
+        # Try to construct a three cycle candidate via a conjugate of t. See
+        # the comment above this function.
+        nrTriedConjugates := nrTriedConjugates + 1;
+        c := t ^ RandomElm(ri, "simplesocle", true)!.el;
+        if not isequal(ri)(t * c, c * t) then
+            nrThreeCycleCandidates := nrThreeCycleCandidates + 1;
+            return (t * c) ^ 2;
+        else
+            # we have to call tryThreeCycleCandidate again
+            return fail;
+        fi;
+    end;
+    # construct the iterator
+    oneThreeCycleCandidate := function()
+        local candidate;
+        repeat
+            if nrInvolutions >= B
+                and (nrTriedConjugates >= C or nrThreeCycleCandidates >= T)
+            then
+                # With probability at least 1 - eps we constructed at least one
+                # three cycle with this iterator.
+                return fail;
+            fi;
+            candidate := tryThreeCycleCandidate();
+            if candidate = NeverApplicable then
+                return NeverApplicable;
+            fi;
+        until candidate <> fail;
+        return candidate;
+    end;
+    return oneThreeCycleCandidate;
+end);
+
+# ri : recog info record with group G
+# c : element of G,
+#     should be a 3-cycle
+# eps : real number, the error bound
+# N : integer, upper bound for the degree of G
+#
+# Returns a list of elements of G.
+#
+# If the input is as assumed, then this function returns a list of bolstering
+# elements with respect to c.
+BindGlobal("BolsteringElements",
+function(ri, c, eps, N)
+    local result, R, S, nrPrebolsteringElms, i, r, cr, cr2;
+    result := [];
+    R := Int(Ceil(7 / 4 * Log2(Float(eps ^ -1))));
+    S := 7 * N * R;
+    nrPrebolsteringElms := 0;
+    i := 0;
+    # find pre-bolstering elements
+    while i <= S and nrPrebolsteringElms <= R do
```

Comment:
Isn't `nrPrebolsteringElms = Length(result)` a loop invariant? Then you could get rid of `nrPrebolsteringElms` and turn this into a for loop:
```suggestion
    for i in [0..S] do
```
and move the `nrPrebolsteringElms <= R` check into the loops...

Reply:
done